File: frontend/src/types/aiFeatures.ts
/**
 * TypeScript types for AI Feature Toggle System
 */

// ============================================
// FEATURE SETTINGS TYPES
// ============================================

export interface AIFeatureSetting {
  id: number;
  feature_id: string;
  feature_name: string;
  description: string | null;
  is_enabled: boolean;
  requires_api_key: boolean;
  api_provider: string | null;
  config: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
  updated_by: number | null;
}

export interface AIFeatureSettingUpdate {
  is_enabled: boolean;
}

// ============================================
// USER PREFERENCE TYPES
// ============================================

export interface UserAIPreference {
  id: number;
  user_id: number;
  feature_id: string;
  is_enabled: boolean;
  admin_override: boolean;
  admin_override_by: number | null;
  created_at: string;
  updated_at: string;
}

export interface UserAIPreferenceUpdate {
  is_enabled: boolean;
}

// ============================================
// FEATURE STATUS TYPES
// ============================================

export interface FeatureStatus {
  feature_id: string;
  feature_name: string;
  description: string | null;
  api_provider: string | null;
  is_enabled: boolean;
  global_enabled: boolean;
  user_enabled: boolean | null;
  admin_override: boolean;
  reason: string;
}

export interface UserFeaturesResponse {
  features: FeatureStatus[];
}

// ============================================
// ADMIN TYPES
// ============================================

export interface FeatureUsageStats {
  total_requests: number;
  total_tokens: number;
  total_cost: number;
  avg_response_time_ms: number;
  success_rate: number;
  period_days: number;
}

export interface AdminFeatureSummary {
  feature_id: string;
  feature_name: string;
  description: string | null;
  is_enabled: boolean;
  api_provider: string | null;
  requires_api_key: boolean;
  enabled_user_count: number;
  total_user_count: number;
  usage_this_month: FeatureUsageStats;
}

export interface AdminFeaturesResponse {
  features: AdminFeatureSummary[];
}

export interface UserPreferenceAdminView {
  user_id: number;
  user_name: string;
  user_email: string;
  preferences: FeatureStatus[];
}

export interface AdminOverrideRequest {
  is_enabled: boolean;
}

// ============================================
// USAGE TYPES
// ============================================

export interface UsageSummary {
  period_days: number;
  total_requests: number;
  total_tokens: number;
  total_cost: number;
  unique_users: number;
  features: Record<string, {
    request_count: number;
    tokens_used: number;
    cost: number;
  }>;
}

export interface UserUsageSummary {
  user_id: number;
  period_days: number;
  total_tokens: number;
  total_cost: number;
  features: Record<string, {
    request_count: number;
    tokens_used: number;
    estimated_cost: number;
  }>;
}

// ============================================
// BATCH OPERATIONS
// ============================================

export interface BatchUserOverrideRequest {
  user_ids: number[];
  feature_id: string;
  is_enabled: boolean;
}

export interface BatchUpdateResponse {
  success: boolean;
  updated_count: number;
  failed_count: number;
  message: string;
}

// ============================================
// FEATURE CHECK RESPONSE
// ============================================

export interface FeatureCheckResponse {
  feature_id: string;
  is_enabled: boolean;
}

// ============================================
// FEATURE ID CONSTANTS
// ============================================

export const AI_FEATURES = {
  SUGGESTIONS: 'ai_suggestions',
  ANOMALY_ALERTS: 'ai_anomaly_alerts',
  PAYROLL_FORECAST: 'ai_payroll_forecast',
  NLP_ENTRY: 'ai_nlp_entry',
  REPORT_SUMMARIES: 'ai_report_summaries',
  TASK_ESTIMATION: 'ai_task_estimation',
} as const;

export type AIFeatureId = typeof AI_FEATURES[keyof typeof AI_FEATURES];

// ============================================
// FEATURE ICONS & COLORS (for UI)
// ============================================

export const FEATURE_UI_CONFIG: Record<string, {
  icon: string;
  color: string;
  bgColor: string;
}> = {
  ai_suggestions: {
    icon: '💡',
    color: 'text-yellow-600',
    bgColor: 'bg-yellow-50',
  },
  ai_anomaly_alerts: {
    icon: '🚨',
    color: 'text-red-600',
    bgColor: 'bg-red-50',
  },
  ai_payroll_forecast: {
    icon: '📈',
    color: 'text-green-600',
    bgColor: 'bg-green-50',
  },
  ai_nlp_entry: {
    icon: '💬',
    color: 'text-blue-600',
    bgColor: 'bg-blue-50',
  },
  ai_report_summaries: {
    icon: '📊',
    color: 'text-purple-600',
    bgColor: 'bg-purple-50',
  },
  ai_task_estimation: {
    icon: '🎯',
    color: 'text-orange-600',
    bgColor: 'bg-orange-50',
  },
};

File: backend/app/schemas/ai_features.py
"""
Pydantic schemas for AI Feature Toggle System.
"""

from datetime import datetime
from decimal import Decimal
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


# ============================================
# GLOBAL FEATURE SETTINGS SCHEMAS
# ============================================

class AIFeatureSettingBase(BaseModel):
    """Base schema for AI feature settings."""
    feature_id: str = Field(..., description="Unique feature identifier")
    feature_name: str = Field(..., description="Display name")
    description: Optional[str] = Field(None, description="Feature description")
    is_enabled: bool = Field(True, description="Global enabled status")
    requires_api_key: bool = Field(True, description="Whether feature requires API key")
    api_provider: Optional[str] = Field(None, description="Required API provider")


class AIFeatureSettingResponse(AIFeatureSettingBase):
    """Schema for returning AI feature settings."""
    id: int
    config: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    updated_by: Optional[int] = None
    
    class Config:
        from_attributes = True


class AIFeatureSettingUpdate(BaseModel):
    """Schema for updating global AI feature settings (admin only)."""
    is_enabled: bool = Field(..., description="Enable or disable the feature globally")


# ============================================
# USER PREFERENCE SCHEMAS
# ============================================

class UserAIPreferenceBase(BaseModel):
    """Base schema for user AI preferences."""
    feature_id: str = Field(..., description="Feature identifier")
    is_enabled: bool = Field(True, description="User's preference for this feature")


class UserAIPreferenceResponse(UserAIPreferenceBase):
    """Schema for returning user AI preferences."""
    id: int
    user_id: int
    admin_override: bool = Field(False, description="Whether admin has overridden this setting")
    admin_override_by: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class UserAIPreferenceUpdate(BaseModel):
    """Schema for updating user's own AI preference."""
    is_enabled: bool = Field(..., description="Enable or disable the feature for yourself")


class AdminOverrideRequest(BaseModel):
    """Schema for admin to override a user's preference."""
    is_enabled: bool = Field(..., description="Force this setting for the user")


# ============================================
# FEATURE STATUS SCHEMAS
# ============================================

class FeatureStatusResponse(BaseModel):
    """Detailed feature status for a user."""
    feature_id: str
    feature_name: str
    description: Optional[str] = None
    api_provider: Optional[str] = None
    is_enabled: bool = Field(..., description="Final computed status")
    global_enabled: bool = Field(..., description="Global admin setting")
    user_enabled: Optional[bool] = Field(None, description="User's preference (if set)")
    admin_override: bool = Field(False, description="Whether admin has overridden")
    reason: str = Field(..., description="Human-readable reason for status")


class UserFeaturesResponse(BaseModel):
    """Complete list of features with status for a user."""
    features: List[FeatureStatusResponse]


# ============================================
# ADMIN SUMMARY SCHEMAS
# ============================================

class FeatureUsageStats(BaseModel):
    """Usage statistics for a feature."""
    total_requests: int = 0
    total_tokens: int = 0
    total_cost: float = 0.0
    avg_response_time_ms: float = 0.0
    success_rate: float = 100.0
    period_days: int = 30


class AdminFeatureSummary(BaseModel):
    """Admin-level feature summary with usage stats."""
    feature_id: str
    feature_name: str
    description: Optional[str] = None
    is_enabled: bool
    api_provider: Optional[str] = None
    requires_api_key: bool = True
    enabled_user_count: int = 0
    total_user_count: int = 0
    usage_this_month: FeatureUsageStats


class AdminFeaturesResponse(BaseModel):
    """Complete admin features summary."""
    features: List[AdminFeatureSummary]


class UserPreferenceAdminView(BaseModel):
    """Admin view of a user's AI preferences."""
    user_id: int
    user_name: str
    user_email: str
    preferences: List[FeatureStatusResponse]


# ============================================
# USAGE LOG SCHEMAS
# ============================================

class AIUsageLogCreate(BaseModel):
    """Schema for creating usage log entry."""
    feature_id: str
    api_provider: Optional[str] = None
    tokens_used: Optional[int] = None
    estimated_cost: Optional[Decimal] = None
    response_time_ms: Optional[int] = None
    success: bool = True
    error_message: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class AIUsageLogResponse(BaseModel):
    """Schema for returning usage log entry."""
    id: int
    user_id: Optional[int] = None
    feature_id: str
    api_provider: Optional[str] = None
    tokens_used: Optional[int] = None
    estimated_cost: Optional[float] = None
    request_timestamp: datetime
    response_time_ms: Optional[int] = None
    success: bool
    error_message: Optional[str] = None
    
    class Config:
        from_attributes = True


class UsageSummaryResponse(BaseModel):
    """Overall AI usage summary."""
    period_days: int
    total_requests: int
    total_tokens: int
    total_cost: float
    unique_users: int
    features: Dict[str, Dict[str, Any]]


class UserUsageSummaryResponse(BaseModel):
    """User-specific usage summary."""
    user_id: int
    period_days: int
    total_tokens: int
    total_cost: float
    features: Dict[str, Dict[str, Any]]


# ============================================
# BATCH OPERATIONS
# ============================================

class BatchUserOverrideRequest(BaseModel):
    """Batch update preferences for multiple users."""
    user_ids: List[int] = Field(..., description="List of user IDs to update")
    feature_id: str = Field(..., description="Feature to update")
    is_enabled: bool = Field(..., description="Enable or disable for these users")


class BatchUpdateResponse(BaseModel):
    """Response for batch operations."""
    success: bool
    updated_count: int
    failed_count: int = 0
    message: str
